test/common/stats/thread_local_store_test.cc

```diff
@@ -90,6 +91,128 @@ class StatsThreadLocalStoreTest : public testing::Test, public RawStatDataAlloca
   std::unique_ptr<ThreadLocalStoreImpl> store_;
 };
 
+class HistogramTest : public testing::Test, public RawStatDataAllocator {
+public:
+  void SetUp() override {
+    InSequence s;
+    ON_CALL(*this, alloc(_)).WillByDefault(Invoke([this](const std::string& name) -> RawStatData* {
+      return alloc_.alloc(name);
+    }));
+
+    ON_CALL(*this, free(_)).WillByDefault(Invoke([this](RawStatData& data) -> void {
+      return alloc_.free(data);
+    }));
+
+    EXPECT_CALL(*this, alloc("stats.overflow"));
+    store_.reset(new ThreadLocalStoreImpl(*this));
+    store_->addSink(sink_);
+    store_->initializeThreading(main_thread_dispatcher_, tls_);
+  }
+
+  void TearDown() override {
+    store_->shutdownThreading();
+    tls_.shutdownThread();
+    // Includes overflow stat.
+    EXPECT_CALL(*this, free(_));
+  }
+
+  std::vector<uint64_t> h1_cumulative_values, h2_cumulative_values, h1_interval_values,
+      h2_interval_values;
+
+  typedef std::map<std::string, Stats::ParentHistogramSharedPtr> NameHistogramMap;
+
+  NameHistogramMap makeHistogramMap(const std::list<ParentHistogramSharedPtr>& hist_list) {
+    NameHistogramMap name_histogram_map;
+    for (const Stats::ParentHistogramSharedPtr& histogram : hist_list) {
+      // Exclude the scope part of the name.
+      const std::vector<std::string>& split_vector = absl::StrSplit(histogram->name(), '.');
+      name_histogram_map.insert(std::make_pair(split_vector.back(), histogram));
+    }
+    return name_histogram_map;
+  }
+
+  /**
+   * Validates that Histogram merge happens as desired and returns the processed histogram count
+   * that can be asserted later.
+   */
+  uint64_t validateMerge() {
+    std::atomic<bool> merge_called{false};
```

Comment:
I don't think this needs to be, or should be, atomic.  The done callback is going to be called exactly once.

Whenever I see atomics I go into hyper-paranoia mode because it's easy to create logical races with them, so I'd rather not have them declared as atomic if it's not (necessary and sufficient).

In this case, the race would be if mergeHistograms ran on another thread, then the EXPECT_TRUE would be a logical race independent of whether the data is atomic or not.  In that case, the main remedy I'm aware of to allow this test to block until completion is using a condvar as you did elsewhere.

If this test always completes before mergeHistogram returns, then you don't need the atomic.  If it doesn't, the atomic isn't sufficient.

@mattklein123 I'm wondering if the unit test infrastructure serializes threads and therefore I'm being overly paranoid about wanting to block test progress until they complete. FWIW In the past I've done unit testing with threads in full force, but then had common test infrastructure primitives to help quiesce before expecting results.